sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -717,47 +717,121 @@ gnrc_netif_t *gnrc_netif_get_by_prefix(const ipv6_addr_t *prefix)
 {
-    unsigned idx = UINT_MAX;
+    uint8_t group_data[GNRC_NETIF_L2ADDR_MAXLEN];
+    int res;
 
+    /* can be called out of lock */
+    res = gnrc_netif_ipv6_group_to_l2_group(netif, addr, group_data);
     gnrc_netif_acquire(netif);
+    if (res > 0) {
+
+        res = _netif_ops_set_helper(netif, NETOPT_L2_GROUP,
+                                    group_data, (uint16_t)res);
+        /* link layer does not support multicast, but we can still use
+         * broadcast */
+        if ((res != -ENOTSUP) && (res < 0)) {
+            goto end;
+        }
+    }
+    /* link layer does not support multicast, but we can still use
+     * broadcast */
+    else if (res != -ENOTSUP) {
+        goto end;
+    }
+    res = -ENOMEM;
     for (unsigned i = 0; i < GNRC_NETIF_IPV6_GROUPS_NUMOF; i++) {
         if (ipv6_addr_equal(&netif->ipv6.groups[i], addr)) {
             gnrc_netif_release(netif);
             return i;
         }
-        if ((idx == UINT_MAX) &&
+        if ((res < 0) &&
             (ipv6_addr_is_unspecified(&netif->ipv6.groups[i]))) {
-            idx = i;
+            res = i;
         }
     }
-    if (idx == UINT_MAX) {
-        gnrc_netif_release(netif);
-        return -ENOMEM;
+    if (res < 0) {
+        goto end;
     }
-    memcpy(&netif->ipv6.groups[idx], addr, sizeof(netif->ipv6.groups[idx]));
+    memcpy(&netif->ipv6.groups[res], addr, sizeof(netif->ipv6.groups[res]));
     /* TODO:
      *  - MLD action
      */
+end:
     gnrc_netif_release(netif);
-    return idx;
+    return res;
 }
 
 void gnrc_netif_ipv6_group_leave_internal(gnrc_netif_t *netif,
                                           const ipv6_addr_t *addr)
 {
-    int idx;
+    uint8_t group_data[GNRC_NETIF_L2ADDR_MAXLEN];
+    /* IPv6 addresses that correspond to the same L2 address */
+    unsigned l2_groups = 0;
+    int idx = -1, res;
 
     assert((netif != NULL) && (addr != NULL));
+    /* can be called out of lock */
+    res = gnrc_netif_ipv6_group_to_l2_group(netif, addr, group_data);
+    /* link layer does not support multicast, but might still have used
+     * broadcast */
+    if ((res < 0) && (res != -ENOTSUP)) {
+        return;
+    }
     gnrc_netif_acquire(netif);
-    idx = _group_idx(netif, addr);
+    for (unsigned i = 0; i < GNRC_NETIF_IPV6_GROUPS_NUMOF; i++) {
+        if (res > 0) {
+            uint8_t tmp[GNRC_NETIF_L2ADDR_MAXLEN];
+            if (!ipv6_addr_is_unspecified(&netif->ipv6.groups[i]) &&
+                (gnrc_netif_ipv6_group_to_l2_group(netif,
+                                                   &netif->ipv6.groups[i],
+                                                   tmp) == res)) {
+                if (memcmp(tmp, group_data, res) == 0) {
+                    l2_groups++;
+                }
+            }
+        }
+        if (ipv6_addr_equal(&netif->ipv6.groups[i], addr)) {
+            idx = i;
+        }
+    }
+    if (idx < 0) {
+        goto end;
```

Comment:
This `goto` literally safes one line of code compared to

```C
gnrc_netif_release(netif);
return;
```

but still one has to search for the label to understand the code.